### DISCUSSION TITLE
Inline images extraction

### DISCUSSION THREAD

**Discussion by maiiabocharova** [pymupdf/PyMuPDF#1231] [Q&A] [answered]
There are some small pictures in my PDFs and I am not able to extract them with page.getImageList() (it returns the empty list). Can you please recommend me what I can do to extract those images, maybe another library? I need those images very much)

**Comment by JorjMcKie** [comment]
If this happens, you
1. **_either_** are not looking at images, but at so-called drawings
2. **_or_** are looking at PDF "inline" images.

**_Drawings_** can be extracted via `page.get_drawings()` / `page.get_cdrawings()`. This is a list of dictionaries. Each dict represents a number of interconnected elementary draw commands: lines, curves, rectangles or quads.

Inline images are only contained in the internal page command source (the `/Contents` object(s)) - nowhere else. Method `page.get_text("dict")` (or "rawdict") contains all images on page - inline and others. Each image occurence occurs as many times as it is shown.

**Reply by maiiabocharova** [reply]
Thank you for the reply!
In the page.get_text("dict") I am getting the thing which I need like text "\nd\n~\n\uf093\n", 
page.get_cdrawings() gives me something, but can I make a png or get numpy array out of it? I want to do a template matching afterwards.

**Comment by JorjMcKie** [comment, marked as answer]
`page.get_text("dict")["blocks"]` is a list of blocks on the page. Each one is either a text block or an image block - see documentation for `TextPage`. Image blocks have `block["type"] = 1`. The image binary is contained in `block["image"]`. More info is contained in the other dict keys.

The list of drawings dicts `page.get_drawings()` can be used to re-draw each on some other page - see the docu [here](https://pymupdf.readthedocs.io/en/latest/faq.html#drawing-and-graphics).
Each "path" dict therein has a `path["rect"]`, which is the rectangle containing all the elementary draws in it.
You could also do a `page.get_pixmap(..., clip=path["rect"])` to create an image of the path. Of course there is the risk, that other things (not belonging to the path) are also part of that picture by accident.

**Reply by maiiabocharova** [reply]
I am so sorry to bother you again, but the function 
`[block for block in page.get_text("dict")["blocks"] if block["type"] == 0]`
returns an empty list and when I drawn the drawings on other page the pictograms which I am trying to extract are also not present (other bars are there but those little pictures are not). If you can just take a look...
[1000252_ZCHU21SETGF_UN1219_VGII_NAG.pdf](https://github.com/pymupdf/PyMuPDF/files/7048679/1000252_ZCHU21SETGF_UN1219_VGII_NAG.pdf)

**Reply by JorjMcKie** [reply]
> ... returns an empty list ...

Of course! Because you are selecting text blocks. Image blocks are ``block["type"] == 1``!

**Reply by maiiabocharova** [reply]
Just a typing mistake here, I did it correctly

**Reply by JorjMcKie** [reply]
ok then
which graphics are you interested in: page? description? Things like this?
![grafik](https://user-images.githubusercontent.com/8290722/130841033-3a7d379b-b200-4824-8e5b-af5de1152bb3.png)

**Reply by maiiabocharova** [reply]
Yes, last ones (the pictograms), how have you extracted them? Sorry I just feel so stupid, I spent a lot of time on them and was not able to extract them, in spite of having tried with 3 different libraries.

**Reply by JorjMcKie** [reply]
😎 They are no image whatsoever, they are no drawings, so I tried to select them with my cursor using a PDF viewer, and voilà: they are **_text_**!!!
Then I looked at the used fonts (page = doc[0])
```python
>>> pprint(page.get_fonts())
[(47, 'n/a', 'Type1', 'Times-BoldItalic', 'R10', ''),
 (48, 'n/a', 'Type1', 'Helvetica', 'R11', ''),
 (49, 'ttf', 'TrueType', 'VGAVKN+DRSymb21', 'R12', ''),
 (51, 'n/a', 'Type1', 'Times-Roman', 'R14', ''),
 (45, 'n/a', 'Type1', 'Courier', 'R8', ''),
 (46, 'n/a', 'Type1', 'Times-Italic', 'R9', '')]
```
so these are probably characters from the ``DRSymb21`` font. Therefore I did this:
```python
>>> spans=[]
>>> for b in blocks:
	for l in b["lines"]:
		for s in l["spans"]:
			if s["font"] != "DRSymb21": continue
			spans.append(s)
>>> for i,s in enumerate(spans):
	pix=page.get_pixmap(clip=s["bbox"])
	pix.save("symb-%i.png" % i)
```
This gave me a number of PNGs, some of them the same or nonsense, but the above were among them:
![symb-0](https://user-images.githubusercontent.com/8290722/130844102-02ec9a91-b733-4be8-a9b0-2ab9c5f233c2.png)
![symb-4](https://user-images.githubusercontent.com/8290722/130844120-cd2eb1e1-00b6-4def-b6b5-26d75294b9de.png)

**Reply by maiiabocharova** [reply]
This is so smart, would have never guessed, that there is such a solution!

Thank you!!! Just can't express how I am grateful!

**Reply by maiiabocharova** [reply]
I am so sorry to bother you once again, but it seems that other PDF documents which have the pictograms sometimes they are not not even text:
```
page = pdf_doc[1].get_textpage().extractDICT()

spans = [span for block in page['blocks'] 
         for line in block['lines'] for span in line["spans"]]
sorted(spans, key = lambda x: (x["bbox"][1], x["bbox"][0]))
```
I did this and went through all the spans near the position which I need and the pictogram was not there (In the previous document with which you helped me text corresponding to the span which contained pictogram was "\uf093",'\uf07f' -- so unicode). But for example with this PDF (second page) 
 it's not the case (I tried extracting images -- page.getImageList() extracted the logo of the company. but not the pictogram, drawings and text also have not extracted those pictograms)

This is such a complicated issue...

**Reply by JorjMcKie** [reply]
this time the PDF page contains images, **_257 (!!!)_** in total. Only one is reachable via an xref (xref 9). This is how to show them:
```python
>>> page.clean_contents()  # required for this page: sloppy PDF creator, read the docu
>>> img=page.get_image_info(xrefs=True)
>>> len(img)
257
>>> # going to draw rects around each image, use a shape because of the large number
>>> shape=page.new_shape()
>>> for i in img:
	shape.draw_rect(i["bbox"])

	
Point(30.479999542236328, 56.64000701904297)
Point(78.95999908447266, 306.7200012207031)
Point(78.72000122070312, 306.96002197265625)
... and so on 257 times ...

>>> shape.finish(color=(0,1,0))  # surround all boxes with green border
>>> shape.commit()  # write shape to page
>>> doc.ez_save("x.pdf")  # save under new name
>>> 
```
Gives you this: 
![grafik](https://user-images.githubusercontent.com/8290722/131108935-ff5de487-31b1-4828-9c80-0fb72cab6282.png)
So except for the logo, all those images are just single 1 or 2 pixels to define your pictograms. An expert at work, obviously!

**Reply by maiiabocharova** [reply]
Thank you so much, so I'll need to look at some other approaches to match pictograms with the templates which I have, definitely those are not to be extracted with standard approaches. 

Thank you for spending time helping me!
